pkg/registers/registers.go

```diff
@@ -57,42 +70,138 @@ func (regs Registers) Find(id RegisterID) Register {
 	return nil
 }
 
-type registerJSON struct {
+type registerForSerializationOBSOLETE struct {
 	ID    RegisterID `json:"id"`
 	Value []byte     `json:"value"`
 }
 
-// MarshalJSON converts a collection of registers into a JSON
-func (regs Registers) MarshalJSON() ([]byte, error) {
-	resultRegisters := make([]registerJSON, 0, len(regs))
-	for _, reg := range regs {
-		v, err := MarshalValue(reg)
+type registersForSerializationOBSOLETE []registerForSerializationOBSOLETE
+
+func (regs registersForSerializationOBSOLETE) ParseRegisters() (Registers, error) {
+	resultRegisters := make(Registers, 0, len(regs))
+	for idx, regS := range regs {
+		v, err := ValueFromBytes(regS.ID, regS.Value)
+		if err != nil {
+			return nil, fmt.Errorf("unable to un-base64 value %d:'%s': %w", idx, regS.Value, err)
+		}
+		reg, err := New(regS.ID, v)
 		if err != nil {
 			return nil, err
 		}
-		resultRegisters = append(resultRegisters, registerJSON{
+		resultRegisters = append(resultRegisters, reg)
+	}
+	return resultRegisters, nil
+}
+
+func (regs Registers) forSerializationOBSOLETE() (registersForSerializationOBSOLETE, error) {
+	resultRegisters := make(registersForSerializationOBSOLETE, 0, len(regs))
+	for _, reg := range regs {
+		b, err := ValueBytes(reg)
+		if err != nil {
+			return nil, fmt.Errorf("unable to serialize the value of register %T:%v", reg, reg)
+		}
+		resultRegisters = append(resultRegisters, registerForSerializationOBSOLETE{
 			ID:    reg.ID(),
-			Value: v,
+			Value: b,
 		})
 	}
-	return json.Marshal(resultRegisters)
+	return resultRegisters, nil
+}
+
+// MarshalJSON converts a collection of registers into a JSON
+// TODO: remove this, obsolete format
+// OBSOLETE
+func (regs Registers) MarshalJSON() ([]byte, error) {
```

Comment:
JSON itself is not an obsolete format. I am ok to add new.

What if someone has a structure it wants to serialise in json and Registers is a field in that structure?

Reply:
I actually believe we should use a map instead of an array in JSON as well. I kept this `MarshalJSON` only for backward compatibility (to be able to read `registers.json` dumped earlier). Would it be fine if I will do next:
* Will reimplement `MarshalJSON` to save it as map.
* Will implement a backward-compatible parser as a separate helper.

Reply:
Yes, this sounds reasonable! Thanks :)